File: Raven.Client.Java/lib/src/main/java/raven/client/changes/TaskedObservable.java
package raven.client.changes;

import java.io.Closeable;
import java.io.IOException;

import raven.abstractions.closure.Function1;
import raven.client.connection.profiling.ConcurrentSet;


public class TaskedObservable<T> implements IObservable<T> {
  private final LocalConnectionState localConnectionState;
  private final Function1<T, Boolean> filter;
  private final ConcurrentSet<IObserver<T>> subscribers = new ConcurrentSet<>();

  public TaskedObservable(LocalConnectionState localConnectionState, Function1<T, Boolean> filter) {
    this.localConnectionState = localConnectionState;
    this.filter = filter;
  }

  @Override
  public Closeable subscribe(final IObserver<T> observer) {
    localConnectionState.inc();
    subscribers.add(observer);
    return new Closeable() {

      @Override
      public void close() throws IOException {
        localConnectionState.dec();
        subscribers.remove(observer);
      }
    };
  }

  public void send(T msg) {
    try {
      if (!filter.apply(msg)) {
        return;
      }
    } catch (Exception e) {
      error(e);
      return;
    }

    for (IObserver<T> subscriber : subscribers) {
      subscriber.onNext(msg);
    }
  }

  public void error(Exception obj) {
    for (IObserver<T> subscriber : subscribers) {
      subscriber.onError(obj);
    }
  }

}

File: src/Raven.NewClient/Changes/TaskedObservable.cs
using System;
using System.Threading.Tasks;
using Raven.Abstractions.Extensions;
using Raven.NewClient.Database.Util;
using Sparrow.Collections;

namespace Raven.NewClient.Client.Changes
{
    public class TaskedObservable<T, TConnectionState> : IObservableWithTask<T> where TConnectionState : IChangesConnectionState
    {
        private readonly TConnectionState localConnectionState;
        private readonly Func<T, bool> filter;
        private readonly ConcurrentSet<IObserver<T>> subscribers = new ConcurrentSet<IObserver<T>>();

        internal TaskedObservable(
            TConnectionState localConnectionState, 
            Func<T, bool> filter)
        {
            this.localConnectionState = localConnectionState;
            this.filter = filter;
            Task = localConnectionState.Task.ContinueWith(task =>
            {
                task.AssertNotFailed();
                return (IObservable<T>)this;
            });
        }

        public Task<IObservable<T>> Task { get; private set; }

        public IDisposable Subscribe(IObserver<T> observer)
        {
            localConnectionState.Inc();
            subscribers.TryAdd(observer);
            return new DisposableAction(() =>
            {
                localConnectionState.Dec();
                subscribers.TryRemove(observer);
            });
        }

        public void Send(T msg)
        {
            try
            {
                if (filter(msg) == false)
                    return;
            }
            catch (Exception e)
            {
                Error(e);
                return;
            }

            foreach (var subscriber in subscribers)
            {
                subscriber.OnNext(msg);
            }
        }

        public void Error(Exception obj)
        {
            foreach (var subscriber in subscribers)
            {
                subscriber.OnError(obj);
            }
        }
    }
}
